### PR TITLE
Parse units in filesizes in Media RSS

### PR DESCRIPTION
**Pull request by sagen**
Had issues with the fileSizes-attribute while parsing a bunch of media rss feeds. It seems like some specify MB/GB as part of filesize. Which is not really compatible with Media RSS-specification (http://www.rssboard.org/media-rss) which specifies fileSize as "... the number of bytes of the media object. It is an optional attribute."

But this modification made rome more useful for me. So it depends on how badly you want to enforce the standard i guess :)

### CODE REVIEW

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java [hunk 1]
```diff
@@ -92,6 +81,30 @@ public Module parse(final Element mmRoot, final Locale locale) {
...
+
+    private long parseFileSize(String fileSize) {
+        Matcher matcher = Pattern.compile("([\\d,.]+)([TGMK])?B", Pattern.CASE_INSENSITIVE).matcher(fileSize);
+        if (matcher.matches()) {
+            BigDecimal number = new BigDecimal(matcher.group(1).replace(',', '.'));
+            BigDecimal multiplier = BigDecimal.valueOf(1);
+            if (matcher.group(2) != null) {
+                char unit  = matcher.group(2).toLowerCase().charAt(0);
+                if (unit == 'k') {
+                    multiplier = BigDecimal.valueOf(1024);
```
Comment (mishako) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Kilobyte is 1000 bytes https://en.wikipedia.org/wiki/Kilobyte
Same of others.

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java [hunk 2]
```diff
@@ -152,7 +165,12 @@ public Module parse(final Element mmRoot, final Locale locale) {
 
                     try {
                         if (content.getAttributeValue("fileSize") != null) {
-                            mc.setFileSize(Long.valueOf(content.getAttributeValue("fileSize")));
+                            String fileSizeAttr = content.getAttributeValue("fileSize").replaceAll("\\s", "");
+                            if (fileSizeAttr.matches("\\d+")) {
```
Comment (mishako) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Maybe handle this case inside `parseFileSize`? For example, return `number` if `matcher.group(2) == null`.

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java [hunk 3]
```diff
@@ -152,7 +165,12 @@ public Module parse(final Element mmRoot, final Locale locale) {
 
                     try {
                         if (content.getAttributeValue("fileSize") != null) {
-                            mc.setFileSize(Long.valueOf(content.getAttributeValue("fileSize")));
+                            String fileSizeAttr = content.getAttributeValue("fileSize").replaceAll("\\s", "");
```
Comment (mishako) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Maybe handle spaces inside `parseFileSize`? For example, include them in the regex. Otherwise we're doing some parsing here and some parsing there.

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java [hunk 4]
```diff
@@ -206,6 +224,7 @@ public Module parse(final Element mmRoot, final Locale locale) {
         return values.toArray(new MediaContent[values.size()]);
     }
 
+
```
Comment (mishako) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Unnecessary blank line.

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java [hunk 5]
```diff
@@ -91,6 +89,30 @@ public void testParse() throws Exception {
         }
     }
 
+
+    public void testParseFileSizeWithUnits() throws Exception {
+        final SyndFeed feed = getSyndFeed("org/rometools/feed/module/mediarss/filesize-with-unit.xml");
```
Comment (mishako) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: Maybe make `parseFileSize` static and package-private? Then you could test it thoroughly here without any xml files. When the method is tested, you could add one test for the integration, i.e. with xml file and `SyndFeed`.

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java [hunk 6]
```diff
@@ -91,6 +89,30 @@ public void testParse() throws Exception {
...
+        final List<SyndEntry> entries = feed.getEntries();
+
+        final MediaEntryModule m1 = (MediaEntryModule) entries.get(0).getModule(MediaModule.URI);
+        assertThat(m1.getMediaContents()[0].getFileSize(), is(valueOf(2).multiply(valueOf(1024).pow(3)).longValue()));
+
+        final MediaEntryModule m2 = (MediaEntryModule) entries.get(1).getModule(MediaModule.URI);
+        assertThat(m2.getMediaContents()[0].getFileSize(), is(valueOf(1.1).multiply(valueOf(1024).pow(3)).longValue()));
```
Comment (mishako) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: It's not obvious what exactly you are testing here. Maybe have a separate test for each assert and use method name to describe what the test is about?

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java [hunk 7]
```diff
@@ -91,6 +89,30 @@ public void testParse() throws Exception {
...
+        assertThat(m3.getMediaContents()[0].getFileSize(), is(valueOf(2).multiply(valueOf(1024).pow(2)).longValue()));
+
+        final MediaEntryModule m4 = (MediaEntryModule) entries.get(3).getModule(MediaModule.URI);
+        assertThat(m4.getMediaContents()[0].getFileSize(), is(2L));
+
+        final MediaEntryModule m5 = (MediaEntryModule) entries.get(4).getModule(MediaModule.URI);
+        assertThat(m5.getMediaContents()[0].getFileSize(), is(valueOf(5.5).multiply(valueOf(1024).pow(4)).longValue()));
+
+        final MediaEntryModule m6 = (MediaEntryModule) entries.get(5).getModule(MediaModule.URI);
+        assertThat(m6.getMediaContents()[0].getFileSize(), is(valueOf(9000).multiply(valueOf(1024)).longValue()));
```
Comment (mishako) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: Why 9000? Is this test about the number or the unit? Same for `5,5TB` and others.

**Comment by mishako** [comment]
Hei Alexander, takk for bidraget! We definitely want Rome to be lenient. While enforcing the standard is important, I think helping our users is more important. Please check my comments and change the code if you agree with them.

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java (hunk 1 above)
Comment (sagen) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: So it seems. Thought i did some proper sampling on this earlier, but some more proper testing now indicates 1000 being the sane choice even though I've seen both being used in the wild. What SI/IEC defines as a KB and KiB doesn't really seem relevant though, as file sizes historically has used 1024 as KB while transfer speeds as been multiples of 1000. This seems like an edge case :)

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java (hunk 2 above)
Comment (sagen) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Yup, agreed. Tried to make as little change as possible to how all the attributes are parsed inline in the method, but this makes more sense. Can't check if matcher.group(2) is null though, because that'd require a regex match, which isn't the case for pure numeric values. So did a check for \d+ as a special first case in parseFileSize.

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java (hunk 3 above)
Comment (sagen) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: Moved to parseFileSize. Reckon the regex is complicated enough as it is though, and it's nice to be able to handle the case of whitespace in an otherwise numeric attribute value, so I left it as a seperate step in parseFileSize

File: rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java (hunk 4 above)
Comment (sagen) [inline rome-modules/src/main/java/com/rometools/modules/mediarss/io/MediaModuleParser.java]: removed

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java (hunk 5 above)
Comment (sagen) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: Did exactly this in the last commit. This test became an really ugly variant of me wanting to copy how it's done elsewhere in the project with multiple asserts and wanting to test some edge cases due to the inherent implicit complexity of regexes :p

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java (hunk 6 above)
Comment (sagen) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: Yes, see previous comment

File: rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java (hunk 7 above)
Comment (sagen) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java]: Uhm, just trying to test that 9000KB is parsed properly even though it should've been 9MB. Never mind, new tests :)

**Comment by sagen** [comment]
Good stuff! Glad to help out on a really useful library :) And thanks for the thorough code review. I've replied to your comments and updated the PR with changes accordingly.

File: rome-modules/src/test/java/com/rometools/modules/mediarss/io/MediaModuleParserTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
+package com.rometools.modules.mediarss.io;
```
Comment (mishako) [inline rome-modules/src/test/java/com/rometools/modules/mediarss/io/MediaModuleParserTest.java]: Please add a license header, like here: https://github.com/rometools/rome/blob/master/rome-utils/src/main/java/com/rometools/utils/Strings.java

**Comment by mishako** [comment]
Nice! Now what's left is the license header and to squash commits. Apart from that :+1:

**Comment by sagen** [comment]
Done and done

**Comment by mishako** [comment]
We don't have automatic builds for pull requests yet, so I ran `mvn verify` locally:
```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] rome/rome-modules/src/test/java/com/rometools/modules/mediarss/MediaModuleTest.java:[99,43] cannot find symbol
  symbol:   class BigDecimal
  location: class com.rometools.modules.mediarss.MediaModuleTest
```

**Comment by sagen** [comment]
Ops, removed some irrelevant import reordering and that import got lost. fixed.
